ir/inst.h

```diff
@@ -3,23 +3,70 @@
 
 #include <vector>
 
-#include "types.h"
+#include "vreg.h"
 #include "opcodes.h"
 
-struct Inst
+class Function;
+
+class Inst
 {
-    enum Opcode {
-#define INST_TYPES(op, ...) op,
-    OPCODES_LIST(INST_TYPES)
-#undef INST_TYPES
-        INVALID
-    };
-
-    Opcode opcode = Opcode::INVALID;
-
-    std::vector<size_t> vregs;
-    std::vector<IdType> ids;
-    std::vector<ImmType> imms;
+public:
+    // maybe just a vector of VReg* as input?
+    template <typename... input>
+    static Inst* InstBuilder(Opcode opcode, input... inputs) {
+        return new Inst(opcode, std::vector<VReg*>{inputs...});
+    }
+
+    ACCESSOR_MUTATOR(opcode_, Opcode, Opcode)
+    ACCESSOR_MUTATOR(inst_id_, InstId, uint32_t)
+    ACCESSOR_MUTATOR(owner_, Function, Function*)
+
+    VReg* GetInput(uint32_t index) const
+    {
+        assert(index < inputs_.size());
+        return inputs_[index];
+    }
+
+    uint32_t GetInputsSize() const
+    {
+        return inputs_.size();
+    }
+
+    const std::vector<VReg*>& GetInputs() const
+    {
+        return inputs_;
+    }
+
+    void Dump();
+
+private:
+    Inst(Opcode opcode, std::vector<VReg*> inputs) : opcode_(opcode), inputs_(inputs) {
+        ++max_inst_id_;
+        inst_id_ = max_inst_id_;
+    }
+
+    void DumpOpcode()
```

Comment:
Can be implemented by creating a `constexpr std::array<const char *, N> names{...}` and getting the name with `names[static_cast<int>(opcode)]`

Reply:
Yes, sure. But I suggest to do it in further versions